packages/block-library/src/heading/autogenerate-anchors.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * External dependencies
+ */
+import { isNil, deburr, trim, startsWith } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { cleanForSlug } from '@wordpress/url';
+import { useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+
+/**
+ * Runs a callback over all blocks, including nested blocks.
+ *
+ * @param {Object[]} blocks   The blocks.
+ * @param {Function} callback The callback.
+ *
+ * @return {void}
+ */
+const recurseOverBlocks = ( blocks, callback ) => {
+	for ( const block of blocks ) {
+		// eslint-disable-next-line callback-return
+		callback( block );
+		if ( block.innerBlocks ) {
+			recurseOverBlocks( block.innerBlocks, callback );
+		}
+	}
+};
+
+/**
+ * Returns the text without markup.
+ *
+ * @param {string} text The text.
+ *
+ * @return {string} The text without markup.
+ */
+const getTextWithoutMarkup = ( text ) => {
+	const dummyElement = document.createElement( 'div' );
+	dummyElement.innerHTML = text;
+	return dummyElement.innerText;
+};
+
+/**
+ * Get all heading anchors.
+ *
+ * @param {Object} blockList An object containing all blocks.
+ * @param {string} excludeId A block ID we want to exclude.
+ *
+ * @return {string[]} Return an array of anchors.
+ */
+const getAllHeadingAnchors = ( blockList, excludeId ) => {
+	const anchors = [];
+
+	recurseOverBlocks( blockList, ( block ) => {
+		if (
+			block.name === 'core/heading' &&
+			( ! excludeId || block.clientId !== excludeId ) &&
+			block.attributes.anchor
+		) {
+			anchors.push( block.attributes.anchor );
+		}
+	} );
+
+	return anchors;
+};
+
+/**
+ * Generate the anchor for a heading.
+ *
+ * @param {string}   anchor            The heading anchor.
+ * @param {string}   content           The block content.
+ * @param {string[]} allHeadingAnchors An array containing all headings anchors.
+ *
+ * @return {string|null} Return the heading anchor.
+ */
+const generateAnchor = ( anchor, content, allHeadingAnchors ) => {
+	content = getTextWithoutMarkup( content );
+
+	// Get the slug.
+	let slug = cleanForSlug( content );
```

Comment:
I believe it would be better to extract this logic to a  `getSlug` function as it would be clearer and easier to test. Also a couple of tests for this (`latin/non-latin`) cases would be good.